{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "14f25906-3f54-4396-9b07-92abda2427f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----------------------+-----------------------------------------+----------------------------------+-------------+-----------------+---------+\n",
      "|        LOCUS         |                  SOURCE                 |               TAXO               | GENOME SIZE | NUMBER OF GENES |   GC%   |\n",
      "+----------------------+-----------------------------------------+----------------------------------+-------------+-----------------+---------+\n",
      "|      NC_022790       |           Marine gokushovirus           |          Gokushovirinae          |     4129    |        6        | 39.57 % |\n",
      "|      NC_002180       |    Chlamydia pneumoniae phage CPAR39    |       Chlamydiamicrovirus        |     4532    |        6        | 40.62 % |\n",
      "|      NC_026013       |           Microviridae IME-16           |    unclassified Microviridae     |     5755    |        7        | 37.06 % |\n",
      "|      NC_001422       | Enterobacteria phage phiX174 sensu lato |            Microvirus            |     5386    |        11       | 44.76 % |\n",
      "|      NC_007821       |   Enterobacteria phage WA13 sensu lato  |            Microvirus            |     6068    |        10       | 44.82 % |\n",
      "|      NC_007856       |   Enterobacteria phage ID18 sensu lato  |            Microvirus            |     5486    |        11       | 45.22 % |\n",
      "|      NC_001420       |    Enterobacteria phage G4 sensu lato   |            Microvirus            |     5577    |        11       | 45.68 % |\n",
      "|      NC_003438       |        Spiroplasma phage 4 (Sp-4)       |         Spiromicrovirus          |     4421    |        9        | 32.05 % |\n",
      "|      NC_002194       |            Chlamydia phage 2            |       Chlamydiamicrovirus        |     4563    |        8        | 40.93 % |\n",
      "|      NC_001998       |        Guinea pig Chlamydia phage       |       Chlamydiamicrovirus        |     4529    |        9        | 40.98 % |\n",
      "|      NC_001741       |  Chlamydia phage 1 (Bacteriophage chp1) |       Chlamydiamicrovirus        |     4877    |        12       | 36.49 % |\n",
      "|      NC_026665       |       Eel River basin pequenovirus      |    unclassified Microviridae     |     6083    |        9        | 30.65 % |\n",
      "|      NC_027633       |        Gokushovirinae Bog1183_53        |          Gokushovirinae          |     4609    |        5        | 36.53 % |\n",
      "|      NC_027634       |         Microviridae Bog1249_12         |    unclassified Microviridae     |     4451    |        6        | 42.37 % |\n",
      "|      NC_027635       |         Microviridae Bog5275_51         |    unclassified Microviridae     |     4815    |        5        | 47.62 % |\n",
      "|      NC_027636       |        Gokushovirinae Bog5712_52        |          Gokushovirinae          |     4501    |        6        | 51.74 % |\n",
      "|      NC_027637       |        Gokushovirinae Bog8989_22        |          Gokushovirinae          |     4656    |        6        | 39.79 % |\n",
      "|      NC_027638       |         Microviridae Bog9017_22         |    unclassified Microviridae     |     4671    |        4        | 41.93 % |\n",
      "|      NC_027639       |         Microviridae Fen2266_11         |    unclassified Microviridae     |     4366    |        5        | 40.60 % |\n",
      "|      NC_027640       |          Microviridae Fen418_41         |    unclassified Microviridae     |     4949    |        5        | 42.79 % |\n",
      "|      NC_027641       |         Microviridae Fen4707_41         |    unclassified Microviridae     |     4552    |        6        | 44.46 % |\n",
      "|      NC_027642       |         Gokushovirinae Fen672_31        |          Gokushovirinae          |     4624    |        7        | 49.63 % |\n",
      "|      NC_027643       |          Microviridae Fen685_11         |    unclassified Microviridae     |     4654    |        5        | 45.98 % |\n",
      "|      NC_027644       |         Microviridae Fen7786_21         |    unclassified Microviridae     |     4686    |        5        | 43.06 % |\n",
      "|      NC_027645       |        Gokushovirinae Fen7875_21        |          Gokushovirinae          |     4631    |        6        | 53.44 % |\n",
      "|      NC_027646       |         Microviridae Fen7895_21         |    unclassified Microviridae     |     4424    |        5        | 43.71 % |\n",
      "|      NC_027647       |         Microviridae Fen7918_21         |    unclassified Microviridae     |     4199    |        5        | 44.53 % |\n",
      "|      NC_027648       |         Microviridae Fen7940_21         |    unclassified Microviridae     |     4453    |        5        | 52.77 % |\n",
      "|      NC_002643       |        Bdellovibrio phage phiMH2K       |         Bdellomicrovirus         |     4594    |        11       | 46.10 % |\n",
      "|      NC_008355       |            Chlamydia phage 3            | unclassified Chlamydiamicrovirus |     4554    |        8        | 41.04 % |\n",
      "|      NC_001330       |       Enterobacteria phage alpha3       |            Microvirus            |     6087    |        10       | 45.17 % |\n",
      "|      NC_007461       |            Chlamydia phage 4            | unclassified Chlamydiamicrovirus |     4530    |        8        | 41.16 % |\n",
      "|      NC_007817       | Enterobacteria phage ID2 Moscow/ID/2001 |  Enterobacteria phage G4 clade   |     5486    |        11       | 45.75 % |\n",
      "|      NC_012868       |        Enterobacteria phage St-1        |            Microvirus            |     6094    |        11       | 45.20 % |\n",
      "|      NC_015785       |          Microviridae phi-CA82          |    unclassified Microviridae     |     5514    |        10       | 39.71 % |\n",
      "|      NC_021805       |       Cellulophaga phage phi12a:1       |    unclassified Microviridae     |     6478    |        13       | 34.02 % |\n",
      "|      NC_021797       |        Cellulophaga phage phi12:2       |    unclassified Microviridae     |     6453    |        13       | 34.80 % |\n",
      "|     59_Coral_002     |         Uncultured Microviridae         |     Uncultured Microviridae      |     4195    |        5        | 35.13 % |\n",
      "| 66_Microbialites_001 |         Uncultured Microviridae         |     Uncultured Microviridae      |     4436    |        3        | 40.19 % |\n",
      "| 68_Microbialite_003  |         Uncultured Microviridae         |     Uncultured Microviridae      |     4660    |        5        | 41.07 % |\n",
      "|     Bourget_052      |         Uncultured Microviridae         |     Uncultured Microviridae      |     4296    |        6        | 43.90 % |\n",
      "|     Bourget_154      |         Uncultured Microviridae         |     Uncultured Microviridae      |     4219    |        5        | 43.68 % |\n",
      "|     Bourget_164      |         Uncultured Microviridae         |     Uncultured Microviridae      |     4228    |        5        | 40.49 % |\n",
      "|     Bourget_224      |         Uncultured Microviridae         |     Uncultured Microviridae      |     4200    |        5        | 40.69 % |\n",
      "|     Bourget_245      |         Uncultured Microviridae         |     Uncultured Microviridae      |     4165    |        5        | 41.15 % |\n",
      "|     Bourget_248      |         Uncultured Microviridae         |     Uncultured Microviridae      |     4281    |        5        | 40.97 % |\n",
      "|     Bourget_259      |         Uncultured Microviridae         |     Uncultured Microviridae      |     4149    |        5        | 39.86 % |\n",
      "|     Bourget_309      |         Uncultured Microviridae         |     Uncultured Microviridae      |     4197    |        6        | 43.43 % |\n",
      "|     Bourget_332      |         Uncultured Microviridae         |     Uncultured Microviridae      |     4289    |        6        | 41.47 % |\n",
      "|     Bourget_504      |         Uncultured Microviridae         |     Uncultured Microviridae      |     4209    |        5        | 41.12 % |\n",
      "|     Bourget_523      |         Uncultured Microviridae         |     Uncultured Microviridae      |     4193    |        5        | 37.82 % |\n",
      "|     Bourget_915      |         Uncultured Microviridae         |     Uncultured Microviridae      |     4285    |        6        | 42.10 % |\n",
      "|      CF7ML_001       |         Uncultured Microviridae         |     Uncultured Microviridae      |     6723    |        4        | 36.09 % |\n",
      "|  Human_feces_A_013   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5316    |        4        | 48.98 % |\n",
      "|  Human_feces_A_016   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5914    |        7        | 37.23 % |\n",
      "|  Human_feces_A_019   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5663    |        6        | 47.14 % |\n",
      "|  Human_feces_A_020   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5776    |        7        | 46.62 % |\n",
      "|  Human_feces_A_021   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6310    |        5        | 39.23 % |\n",
      "|  Human_feces_A_029   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5158    |        5        | 45.05 % |\n",
      "|  Human_feces_A_032   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5818    |        6        | 40.16 % |\n",
      "|  Human_feces_A_033   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5502    |        4        | 47.60 % |\n",
      "|  Human_feces_A_034   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6301    |        6        | 38.05 % |\n",
      "|  Human_feces_A_047   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5893    |        8        | 35.44 % |\n",
      "|  Human_feces_A_048   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5816    |        8        | 36.39 % |\n",
      "|  Human_feces_B_007   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6233    |        3        | 40.51 % |\n",
      "|  Human_feces_B_020   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6041    |        9        | 38.52 % |\n",
      "|  Human_feces_B_021   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5910    |        8        | 36.27 % |\n",
      "|  Human_feces_B_023   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5733    |        7        | 36.80 % |\n",
      "|  Human_feces_B_029   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5226    |        4        | 48.62 % |\n",
      "|  Human_feces_B_039   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6008    |        4        | 41.04 % |\n",
      "|  Human_feces_B_068   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5123    |        5        | 41.36 % |\n",
      "|  Human_feces_C_010   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6364    |        5        | 37.82 % |\n",
      "|  Human_feces_C_014   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5536    |        5        | 46.76 % |\n",
      "|  Human_feces_C_016   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5985    |        4        | 41.31 % |\n",
      "|  Human_feces_C_029   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5888    |        7        | 36.44 % |\n",
      "|  Human_feces_C_031   |         Uncultured Microviridae         |     Uncultured Microviridae      |     4810    |        3        | 45.03 % |\n",
      "|  Human_feces_C_043   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6226    |        5        | 37.79 % |\n",
      "|  Human_feces_D_008   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6582    |        4        | 39.53 % |\n",
      "|  Human_feces_D_014   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5274    |        4        | 48.78 % |\n",
      "|  Human_feces_D_015   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6231    |        3        | 38.93 % |\n",
      "|  Human_feces_D_022   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5481    |        4        | 47.96 % |\n",
      "|  Human_feces_D_031   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6659    |        6        | 39.93 % |\n",
      "|  Human_feces_D_045   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5299    |        4        | 44.97 % |\n",
      "|  Human_feces_E_007   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5774    |        7        | 45.16 % |\n",
      "|  Human_feces_E_009   |         Uncultured Microviridae         |     Uncultured Microviridae      |     4928    |        4        | 43.12 % |\n",
      "|  Human_feces_E_010   |         Uncultured Microviridae         |     Uncultured Microviridae      |     4881    |        4        | 44.58 % |\n",
      "|  Human_feces_E_011   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6313    |        3        | 36.89 % |\n",
      "|  Human_feces_E_017   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5769    |        7        | 45.96 % |\n",
      "|   Human_gut_21_005   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6399    |        6        | 37.64 % |\n",
      "|   Human_gut_21_019   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5029    |        6        | 40.84 % |\n",
      "|   Human_gut_22_017   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6251    |        5        | 39.65 % |\n",
      "|   Human_gut_24_085   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5281    |        5        | 32.56 % |\n",
      "|   Human_gut_27_015   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5035    |        7        | 40.97 % |\n",
      "|   Human_gut_27_035   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5266    |        5        | 45.59 % |\n",
      "|   Human_gut_30_017   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5755    |        4        | 41.80 % |\n",
      "|   Human_gut_30_040   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6061    |        6        | 35.12 % |\n",
      "|   Human_gut_31_037   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6166    |        10       | 37.22 % |\n",
      "|   Human_gut_31_045   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5266    |        5        | 45.59 % |\n",
      "|   Human_gut_31_054   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5036    |        6        | 40.98 % |\n",
      "|   Human_gut_31_126   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5696    |        4        | 44.48 % |\n",
      "|   Human_gut_32_012   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5695    |        4        | 44.49 % |\n",
      "|   Human_gut_32_015   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5279    |        5        | 32.43 % |\n",
      "|   Human_gut_32_030   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5266    |        5        | 45.59 % |\n",
      "|   Human_gut_33_003   |         Uncultured Microviridae         |     Uncultured Microviridae      |     5704    |        6        | 43.68 % |\n",
      "|   Human_gut_33_005   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6449    |        5        | 40.14 % |\n",
      "|   Human_gut_33_017   |         Uncultured Microviridae         |     Uncultured Microviridae      |     6171    |        5        | 40.20 % |\n",
      "|   Human_gut_33_018   |         Uncultured Microviridae         |     Uncultured Microviridae      |     4799    |        5        | 43.34 % |\n",
      "|   Human_gut_33_023   |         Uncultured Microviridae         |     Uncultured Microviridae      |     4549    |        5        | 40.49 % |\n",
      "|   Human_gut_34_012   |         Uncultured Microviridae         |     Uncultured Microviridae      |     4822    |        5        | 43.11 % |\n",
      "|   Human_gut_35_025   |         Uncultured Microviridae         |     Uncultured Microviridae      |     4822    |        5        | 43.11 % |\n",
      "|   Human_gut_36_019   |         Uncultured Microviridae         |     Uncultured Microviridae      |     4822    |        5        | 43.07 % |\n",
      "|   Human_gut_37_015   |         Uncultured Microviridae         |     Uncultured Microviridae      |     4822    |        5        | 43.09 % |\n",
      "|       JCVI_001       |         Uncultured Microviridae         |     Uncultured Microviridae      |     4125    |        5        | 47.00 % |\n",
      "|       JCVI_003       |         Uncultured Microviridae         |     Uncultured Microviridae      |     4240    |        5        | 43.46 % |\n",
      "|      Pavin_110       |         Uncultured Microviridae         |     Uncultured Microviridae      |     4194    |        6        | 41.91 % |\n",
      "|      Pavin_279       |         Uncultured Microviridae         |     Uncultured Microviridae      |     3992    |        5        | 39.65 % |\n",
      "|      Pavin_723       |         Uncultured Microviridae         |     Uncultured Microviridae      |     3989    |        4        | 31.33 % |\n",
      "|   SectLung2LLL_002   |         Uncultured Microviridae         |     Uncultured Microviridae      |     4196    |        5        | 39.27 % |\n",
      "+----------------------+-----------------------------------------+----------------------------------+-------------+-----------------+---------+\n"
     ]
    }
   ],
   "source": [
    "# %load microviridae.py\n",
    "#!/usr/bin/python3\n",
    "\n",
    "from collections import defaultdict\n",
    "from prettytable import PrettyTable\n",
    "\n",
    "##################################################################################################################################\n",
    "# This script allows for the extraction of information from Microviridae from two files, \"microviridae.gbf\" and \"proteine_micro.faa\n",
    "##################################################################################################################################\n",
    "\n",
    "locus=[]; source=[]; taxo=[]; leng=[]; numGene=[]; prCG=[]\n",
    "d_locus_genome=defaultdict(str); d_locus_prCG=defaultdict(str); d_locus_nbGene=defaultdict(int);  \n",
    "orga=False; genome=False\n",
    "\n",
    "with open(\"microviridae.gbf\") as f1 :\n",
    "\tfor li in f1 :\n",
    "\t\tli = li.rstrip(\"\\n\")\n",
    "\n",
    "\t\tif li.startswith(\"LOCUS\"):\n",
    "\t\t\tlocu=li.split()[1]\n",
    "\t\t\tleng.append(li.split()[2])  # liste contient le nombre des genes\n",
    "\t\t\tlocus.append(li.split()[1]) # liste contient tous les locus\n",
    "\n",
    "\t\tif li.startswith(\"SOURCE\"):\n",
    "\t\t\tli = \" \".join(li.split()[1:])\n",
    "\t\t\tsource.append(li)           # liste contient les source de chaque locus\n",
    "\n",
    "\t\tif li.startswith('ORIGIN'):\n",
    "\t\t\tgenome=True\n",
    "\n",
    "\t\tif li.startswith('LOCUS'):\n",
    "\t\t\tgenome=False\n",
    "\n",
    "\t\tif genome== True :\n",
    "\t\t\tligen=li.rstrip('\\n').split()[1:]\n",
    "\t\t\tligen=\"\".join(ligen)\n",
    "\t\t\td_locus_genome[locu]+=ligen.upper() # Creating a dictionary containing loci as keys, and the genome as values, with each nucleotide in uppercase\n",
    "\n",
    "\t\tif li.startswith(\"  ORGANISM\"):\n",
    "\t\t\torga=True\n",
    "\n",
    "\t\tif li.startswith(\"REFERENCE \"):\n",
    "\t\t\torga=False\n",
    "\n",
    "\t\tif orga==True :\n",
    "\t\t\tif not li.startswith(\"  ORGANISM \") and li.endswith(\".\"):\n",
    "\t\t\t\ttaxo.append(li.split(\";\")[-1].strip().strip(\".\").replace(\"environmental samples\", \"Uncultured Microviridae\")) # liste taxo qui contient la taxonomie de chaque locus\n",
    "\n",
    "# Creating a dictionary with the key \"locus\" and the value \"percentage of CG.\"\n",
    "\n",
    "for i in d_locus_genome:\n",
    "\tcg=str((d_locus_genome[i].count(\"C\")+d_locus_genome[i].count(\"G\"))/len(d_locus_genome[i])*100)[:5]+\" %\"\n",
    "\td_locus_prCG[i]=cg\n",
    "\n",
    "\n",
    "\n",
    "with open(\"proteine_micro.faa\") as f2 :\n",
    "\tfor li in f2 :\n",
    "\t\tli = li.rstrip(\"\\n\")\n",
    "\t\tif li.startswith(\">\"):\n",
    "\t\t\td_locus_nbGene[li.split(\"###\")[1].strip()]+=1     # The dictionary contains 'locus' as the key and 'number of genes' as the value\n",
    "\n",
    "# Creating lists from dictionary values to zip them together to obtain an array of all Microviridae viruses\n",
    "\n",
    "for i in locus :\n",
    "\tfor j in d_locus_nbGene:\n",
    "\t\tif i == j :\n",
    "\t\t\tnumGene.append(d_locus_nbGene[j])       # The list contains the number of genes for each locus\n",
    "\n",
    "\n",
    "for i in locus :\n",
    "\tfor h in d_locus_prCG:\n",
    "\t\tif i == h:\n",
    "\t\t\tprCG.append(d_locus_prCG[i])    # The list contains the percentage of each locus \n",
    "\n",
    "# \"The final result of the 118 genomes is the creation of a table.\"\n",
    "\n",
    "x = PrettyTable()\n",
    "x.field_names = [\"LOCUS\", \"SOURCE\", \"TAXO\", \"GENOME SIZE\", \"NUMBER OF GENES\", \"GC%\"]       #Table Header \n",
    "\n",
    "for a, b, c, d, e, f in zip(locus, source, taxo, leng, numGene, prCG):\n",
    "\tx.add_row([a, b, c, d, e, f])                                                       # remplir notre tableau\n",
    "print(x)\n",
    "\n",
    "\t\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa3416ec-52bc-4318-bf89-da3cd329b4a3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
